Current week

Edwu29 opened the pull request:
## Reference Issues
Closes #121 
## Summary of Change/Fix 
Migrates getWeek function from PeterPortal Client to PeterPortal API. Fetches from reg.uci.edu, **does not use mongo**. 

- week.helper.js contains the helper functions. Main function is getWeek()
- REST endpoint /schedule/week to get data.
- Arguments include year, month, day. Must include all or none.
- GraphQL type is created for weekType and is queried.

## Test Plan
Run ``` npm test ```
Tested REST and GQL queries
### REST:
#### Valid:
All arguments: http://localhost:8080/rest/v0/schedule/week?year=2022&month=03&day=29
No arguments: http://localhost:8080/rest/v0/schedule/week
All arguments: http://localhost:8080/rest/v0/schedule/week?year=2015&month=02&day=29
All arguments: http://localhost:8080/rest/v0/schedule/week?year=2015&month=03&day=29
#### Invalid:
Invalid day: http://localhost:8080/rest/v0/schedule/week?year=2022&month=02&day=299
Invalid month: http://localhost:8080/rest/v0/schedule/week?year=2022&month=042&day=29
Year invalid: http://localhost:8080/rest/v0/schedule/week?year=202233&month=02&day=28
Missing arguments: http://localhost:8080/rest/v0/schedule/week?year=2022&month=02
Invalid month & day: http://localhost:8080/rest/v0/schedule/week?year=2022&month=023&day=299

### GQL:
#### valid:
```
arguments valid
{
  week(year:"2022", month: "02", day: "29") {
    week
    quarter
    display
  }
}
```
```
no arguments -> give current week
{
  week {
    week
    quarter
    display
  }
}
```
```
valid arguments
{
  week(year:"2015", month: "02", day: "28") {
    week
    quarter
    display
  }
}
```
```
only week and quarter requested
{
  week(year:"2022", month: "02", day: "28") {
    week
    quarter
  }
}
```
#### Invalid:
```
invalid day
{
  week(year:"2022", month: "02", day: "299") {
    week
    quarter
    display
  }
}
```
```
invalid month
{
  week(year:"2022", month: "042", day: "29") {
    week
    quarter
    display
  }
}
```
```
missing day argument
{
  week(year:"2022", month: "02") {
    week
    quarter
    display
  }
}
```

ramanxg commented on graphql/schema.js:
```diff
@@ -251,6 +255,16 @@ const courseOfferingType = new GraphQLObjectType({
+const weekType = new GraphQLObjectType({
+  name: 'Week',
+  fields: () => ({
+    week: { type: GraphQLInt },
+    quarter: { type: GraphQLString },
+    display: { type: GraphQLString }
```
Add descriptions for these argument fields.

Edwu29 replied:
Agreed

ramanxg reviewed (changes requested):
When graphql, throws an error, it doesn't give a message on invalid date arguments. It should give a message to tell users what is wrong.

Edwu29 commented on package.json:
```diff
@@ -28,6 +29,7 @@
     "graphql-playground-middleware-express": "^1.7.21",
     "http-errors": "~1.6.3",
     "moesif-aws-lambda": "^1.3.0",
+    "mongodb": "^4.4.0",
```
Right

Edwu29 commented on graphql/schema.js:
```diff
@@ -18,6 +19,9 @@ var {getAllInstructors, getInstructor, getUCINetIDFromName} = require('../helper
 const { ValidationError } = require('../helpers/errors.helper');
+var {createErrorJSON} = require("../helpers/errors.helper")
+const { getWeek } = require('../helpers/week.helper');
+const { json } = require('express');
```
Right as well

Edwu29 commented on graphql/schema.js:
```diff
@@ -437,6 +451,26 @@ const queryType = new GraphQLObjectType({
+         return val
+      //  }).catch( err => {
+      //   return createErrorJSON(400, "Bad Request: Invalid parameter", "Unable to complete websoc-api query");
+       })
```
Agreed

Edwu29 commented:
Updates invalid message for both REST and GQL, resolves promises. Uses async/await

Edwu29 commented:
`Year invalid, but considered valid: http://localhost:8080/rest/v0/schedule/week?year=202233&month=02&day=28`
This isn't considered invalid by JavaScript date object for some reason..

Also, can create new issues for `week` documentation and automated tests.

ramanxg commented:
> `Year invalid, but considered valid: http://localhost:8080/rest/v0/schedule/week?year=202233&month=02&day=28`
> This isn't considered invalid by JavaScript date object for some reason..

I think javascript is just using that year as a the year, so `+202233-02-28T08:00:00.000Z`, this is the date string and 202233 is the year. We should add our own validation. Just check that it's 4 character string, We can add in the documentation what years are available.

Also, for very early years, like 2009, 2008, it's broken.

http://localhost:8080/rest/v0/schedule/week?year=2008&month=01&day=30
It will make a request to https://reg.uci.edu/calendars/quarterly/2008-2009/quarterly8-9.html

but it needs to be `quarterly08-09.html`

Edwu29 commented:
- Added validation for year, month and day strings. Has to be YYYY, MM, DD with length 4, 2, 2 accordingly.
- Fixed url request for years 2008 and 2009

Note: UCI reg only offers info from Fall 2008 (Sept 25) - Spring 2023 at the moment.

ramanxg reviewed (changes requested):
Looks good, just one small change I requested, and fix the merge conflicts with master.

Edwu29 commented:
Fixed issues and Test Plan is successful

Edwu29 commented:
Dates in JavaScript reference- https://css-tricks.com/everything-you-need-to-know-about-date-in-javascript/
